crates/bifrost/src/providers/replicated_loglet/tasks/find_tail.rs

```diff
@@ -0,0 +1,645 @@
+                        // repair [known_global_tail..max-tail]?
+                        nodeset_checker.check_write_quorum(|attribute| match attribute {
+                            // We have all copies of the max-local-tail replicated. It's safe to
+                            // consider this offset as committed.
+                            NodeTailStatus::Known { local_tail, sealed } => {
+                                *sealed && (*local_tail >= max_local_tail)
+                            }
+                            _ => false,
+                        }) {
+                                // Great. All nodes sealed and we have a stable value to return.
+                                //
+                                // todo: If some nodes have lower global-tail than max-local-tail, then
+                                // broadcast a a release to max-tail to avoid unnecessary repair if
+                                // underreplication happened after this point.
+                                inflight_info_requests.abort_all();
+                                // Note: We don't set the known_global_tail watch to this value nor the
+                                // seal bit. The caller can decide whether to do that or not based on the
+                                // use-case.
+                                return FindTailResult::Sealed {
+                                    global_tail: max_local_tail,
+                                };
+                            } else {
+                                // F-majority sealed, but tail needs repair in range
+                                // [current_known_tail..max_local_tail]
+                                //
+                                // todo: Although we have f-majority, it's not always guaranteed that we are
+                                // able to form a write-quorum within this set of nodes. For instance, if
+                                // replication-factor is 4 in a nodeset of 5 nodes, F-majority is 2 nodes which
+                                // isn't enough to form write-quorum. In this case, we need to wait for more
+                                // nodes to chime in. But we _only_ need this if max_local_tail is higher than
+                                // current_known_global. Once we have enough sealed node that match
+                                // write-quorum **and** f-majority, then we can repair the tail.
+                                if nodeset_checker.check_write_quorum(NodeTailStatus::is_known) {
+                                    // We can repair.
+                                    todo!("Tail repair is not implemented yet")
+                                } else {
+                                    // wait for more nodes
+                                    break 'check_nodeset;
+                                }
```

Comment:
Question here about the potential repair process. Will this try to copy the missing logs from the max_local_tail until it forms a write quorum ?

Or will it find the max value in range `[known_global..max_local]` that satisfy the write quorum?

Reply:
Both. It'll first walk backwards from max-tail until it hits a record that's not fully replicated. Then it'll attempt to replicate those records (in the forward direction) until it reaches max-local-tail. It's critical that the order of repair goes in this direction to ensure correctness and repeatability if repair failed or got interrupted.